src/UCTSearch.cpp

```diff
@@ -228,12 +228,25 @@ float UCTSearch::get_min_psa_ratio() const {
     return 0.0f;
 }
 
+class VirtualLossScopeGuard {
+private:
+    UCTNode & m_node;
+public:
+    VirtualLossScopeGuard(UCTNode & n) : m_node(n) {
+        m_node.virtual_loss();
+    }
+    ~VirtualLossScopeGuard() {
+        m_node.virtual_loss_undo();
+    }
+};
+
 SearchResult UCTSearch::play_simulation(GameState & currstate,
                                         UCTNode* const node) {
     const auto color = currstate.get_to_move();
     auto result = SearchResult{};
 
-    node->virtual_loss();
+    // This will undo virtual loss even if something throws an exception
+    VirtualLossScopeGuard scope_guard(*node);
```

Comment:
How about using [Boost.ScopeExit](https://www.boost.org/doc/libs/1_72_0/libs/scope_exit/doc/html/index.html) instead, keeping the code local?